MaterialState migration guide (#10194)
flutter/website: src/release/breaking-changes/material-state.md

---
title: Rename MaterialState to WidgetState
description: >-
  MaterialState and it's related APIs have been moved
  outside of the Material library and renamed to
  WidgetState.
---

## Summary

> `MaterialState`, and its related APIs have been moved out

Reviewer: ```suggestion
`MaterialState`, and its related APIs, have been moved out
```

of the Material library and renamed to `WidgetState`.

## Background

> `MaterialState` provided logic for handeling multiple different

Reviewer: ```suggestion
Previously, `MaterialState` provided logic for handling multiple different
```

states a widget could have, like "hovered", "focused", and

> "disabled". Because this functionality could be useful outside

Reviewer 2: could be useful -> is useful?

> of the Material library, namely the base Widgets layer and

Reviewer: ```suggestion
of the Material library, namely for the base Widgets layer and
```

Cupertino, it was decided to move it outside of Material. As
part of the move, and to avoid future confusion, the different

> MaterialState classes have been renamed to `WidgetState`. The

Reviewer: ```suggestion
`MaterialState` classes have been renamed to `WidgetState`. The
```

behavior of the two are the same.

| Before    | Now |
| -------- | ------- |

> | MaterialState | WidgetState |
> | MaterialStatePropertyResolver | WidgetStatePropertyResolver |
> | MaterialStateColor | WidgetStateColor |
> | MaterialStateMouseCursor | WidgetStateColorMouseCursor |
> | MaterialStateBorderSide | WidgetStateBorderSide |
> | MaterialStateOutlinedBorder | WidgetStateOutlinedBorder |
> | MaterialStateTextStyle | WidgetStateTextStyle |
> | MaterialStateProperty | WidgetStateProperty |
> | MaterialStatePropertyAll | WidgetStatePropertyAll |
> | MaterialStatesController | WidgetStatesController |
> | MaterialStateMixin | WidgetStateMixin |

Reviewer: ```suggestion
| `MaterialState` | `WidgetState` |
| `MaterialStatePropertyResolver` | `WidgetStatePropertyResolver` |
| `MaterialStateColor` | `WidgetStateColor` |
| `MaterialStateMouseCursor` | `WidgetStateColorMouseCursor` |
| `MaterialStateBorderSide` | `WidgetStateBorderSide` |
| `MaterialStateOutlinedBorder` | `WidgetStateOutlinedBorder` |
| `MaterialStateTextStyle` | `WidgetStateTextStyle` |
| `MaterialStateProperty` | `WidgetStateProperty` |
| `MaterialStatePropertyAll` | `WidgetStatePropertyAll` |
| `MaterialStatesController` | `WidgetStatesController` |
| `MaterialStateMixin` | `WidgetStateMixin` |
```

The classes `MaterialStateOutlineInputBorder` and
`MaterialStateUnderlineInputBorder` were left in the

> Material library with no `WidgetState` equivilant, as
> they are Material design specific.

Reviewer: ```suggestion
Material library with no `WidgetState` equivilent, as
they are specific to Material design.
```

## Migration guide

A [Flutter fix][] is available to help migrate the `MaterialState`
classes to `WidgetState`.

To migrate, replace `MaterialState` with `WidgetState`.

Code before migration:

```dart
MaterialState selected = MaterialState.selected;

final MaterialStateProperty<Color> backgroundColor;

class _MouseCursor extends MaterialStateMouseCursor{
    const _MouseCursor(this.resolveCallback);

    final MaterialPropertyResolver<MouseCursor?> resolveCallback;

    @override

>     MouseCursor resolve(Set<MaterialState> states) => resolveCallback(states) ?? MouseCursor.uncontrolled;

Reviewer 2: nit here and below: I think usually we indent by only two spaces.

}

> MaterialStateBorderSide? get side;

Reviewer 2: This top-level getter without a body looks a little strange. Is that even legal in dart? Maybe just have it return null?

```

Code after migration:

> ```dart

Reviewer 2: Same comments about this sample

WidgetState selected = WidgetState.selected;

final WidgetStateProperty<Color> backgroundColor;

class _MouseCursor extends WidgetStateMouseCursor{
    const _MouseCursor(this.resolveCallback);

    final WidgetPropertyResolver<MouseCursor?> resolveCallback;

    @override
    MouseCursor resolve(Set<WidgetState> states) => resolveCallback(states) ?? MouseCursor.uncontrolled;
}

WidgetStateBorderSide? get side;
```

## Timeline

> Landed in version: xxx<br>

Reviewer: @MitchellGoodwin, do you have the info on which main or beta release has this change? Put that here.

Author: It's not landed yet, we are trying to make sure our communication is ready to go before landing it. Should I put what release it's likely to be in for now? Or leave it as a placeholder.

Reviewer: Leave it for now, but once you know, we should update the page.  thx!

---
title: Rename MaterialState to WidgetState
description: >-

>   MaterialState and it's related APIs have been moved

Reviewer 2: ```suggestion
  MaterialState and its related APIs have been moved
```

  outside of the Material library and renamed to
  WidgetState.
---

## Summary

`MaterialState`, and its related APIs, have been moved out
of the Material library and renamed to `WidgetState`.

## Background

Previously, `MaterialState` provided logic for handling multiple different
states a widget could have, like "hovered", "focused", and
"disabled". Because this functionality is useful outside of the
Material library, namely for the base Widgets layer and Cupertino,

> it was decided to move it outside of Material. Aspart of the move, and to

Reviewer 2: ```suggestion
it was decided to move it outside of Material. As part of the move, and to
```
